### PR TITLE
Create systemic mysql pool

### PR DESCRIPTION
**Pull request by servatj**
New pull request for creating and discussing the initial parts of a MySQL systemic component.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    return pool;
+  };
+
+  const start = (dependencies, cb) => {
+    config = dependencies.config || {}
+    logger = dependencies.logger || console
+
+    logger.info('Starting the pool');
+    return initPool(config);
+    cb();
```
Comment (feliun) [inline index.js]: We favour async/await vs callback in latest versions of systemic:

```
async (dependencies) => {
    config = dependencies.config || {}
    logger = dependencies.logger || console
     logger.info('Starting the pool');
    return initPool(config);
}
```

File: index.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  };
+
+  // stop
+
+  const stop = (connectionPool, cb) => {
+    if (connectionPool) {
```
Comment (feliun) [inline index.js]: Same:

```
async (connectionPool) => 
    if (connectionPool) {
      return connectionPool.release();
    }
    return Promise.resolve();
  };
```

File: index.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      return connectionPool.release();
+    }
+    return cb();
+  };
+
+  return {
+    start: start,
```
Comment (feliun) [inline index.js]: ES6:
```
start,
stop
```

File: test/mysql.test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        connectionLimit: 10,
+        host: 'localhost',
+        user: 'admin',
+        password: 'password',
+        database: 'event_aggregates'
+      }
+    };
+    mysql = mysqlComponent(dependencies);
+    pool = await mysql.start((err, component)=> {
+       return component;
```
Comment (feliun) [inline test/mysql.test.js]: err handling please

File: test/mysql.test.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+      config: {
+        connectionLimit: 10,
+        host: 'localhost',
+        user: 'admin',
+        password: 'password',
+        database: 'event_aggregates'
+      }
+    };
+    mysql = mysqlComponent(dependencies);
+    pool = await mysql.start((err, component)=> {
```
Comment (feliun) [inline test/mysql.test.js]: no callback please: `pool = await mysql.start();`

File: test/mysql.test.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+       return component;
+    });
+  })
+
+  after( async () => {
+    mysql.stop(connection)
```
Comment (feliun) [inline test/mysql.test.js]: `await mysql.stop(connection);`

File: test/mysql.test.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  })
+
+  it('It should connect without error' , async () => {
+    connection = await pool.getConnection((err, connection) => connection.release());
```
Comment (feliun) [inline test/mysql.test.js]: this won't work, the await is not waiting for the callback

File: index.js [hunk 8]
```diff
@@ -21,26 +22,21 @@ module.exports = (options) => {
...
+  const start = async (dependencies) => {
     config = dependencies.config || {}
     logger = dependencies.logger || console
-
     logger.info('Starting the pool');
     return initPool(config);
-    cb();
   };
 
   // stop
```
Comment (Betisman) [inline index.js]: Perhaps we could remove this comment

File: index.js [hunk 9]
```diff
@@ -21,26 +22,21 @@ module.exports = (options) => {
...
 
-  const stop = (connectionPool, cb) => {
-    if (connectionPool) {
-      return connectionPool.release();
-    }
-    return cb();
+  const stop = async () => {
```
Comment (Betisman) [inline index.js]: Does this empty function do any difference? Can't we remove it?

File: test/mysql.test.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        connectionLimit: 10,
+        host: 'localhost',
+        user: 'admin',
+        password: 'password',
+        database: 'event_aggregates'
+      }
+    };
+    mysql = mysqlComponent(dependencies);
+    pool = await mysql.start(dependencies.config);
+    mysql.createConnec
```
Comment (Betisman) [inline test/mysql.test.js]: Is this right? Maybe something incomplete?

File: test/mysql.test.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  })
+
+  after( async () => {
+    await mysql.stop();
+  })
+
+  it('It should connect without error' , async () => {
+    connection = await pool.getConnection();
+    connection.release();
+    expect(connection).to.not.be.null;
```
Comment (Betisman) [inline test/mysql.test.js]: Does this work? I thought expect.js used a different assert api:
`expect(connection).to.not.be.(null);`

File: index.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
+const {get} = require('lodash');
```
Comment (Betisman) [inline index.js]: I miss a couple of spaces in the destructuring:
`const { get }...`

File: index.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+      user: config.user || 'admin',
+      password: config.password || 'password',
+      database: config.database || 'event_aggregates'
+    });
+
+    return pool;
+  };
+
+  const start = async (dependencies) => {
+    config = dependencies.config || {}
```
Comment (Betisman) [inline index.js]: In order of consistency, I would close the line using `;`

File: index.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    logger = dependencies.logger || console
```
Comment (Betisman) [inline index.js]: In order of consistency, I would close the line using `;`

File: index.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+  // stop
+
+  const stop = async () => {
+
+  };
+
+  return {
+    start,
+    stop
```
Comment (Betisman) [inline index.js]: In order of consistency, I would close the line using `,`

File: index.js [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+  // configure
+
+  const initPool = async (config) => {
+
+    const pool = await mysql.createPool({
+      connectionLimit: config.ConnectionLimit || 10,
+      host: config.host || 'localhost',
+      user: config.user || 'admin',
+      password: config.password || 'password',
+      database: config.database || 'event_aggregates'
```
Comment (Betisman) [inline index.js]: In order of consistency, I would close the line using `,`

**Review by Betisman — CHANGES_REQUESTED** [review]

File: test/mysql.test.js (hunk 10 above)
Comment (servatj) [reply test/mysql.test.js]: fixed

**Comment by servatj** [comment]
@feliun @Betisman thanks for your review I'm going to merge this as initial systemic component and then add more functionality and things like eslint etc

**Comment by feliun** [comment]
@servatj please bear in mind this should be very lightweight - systemics are just adapters to expose the client from configuration. Not much funcionality should be needed. What did you have in mind?

**Comment by servatj** [comment]
@feliun Ok, I was thinking to do the following stages:

  - Properly read the configuration
  - Validate it, and test it
  - Expose a query function.

What do you think? Or do you think it's better to do similar as we do in systemic-pg and only expose the pool when we start the component?

**Comment by feliun** [comment]
We should enforce config as you said, and then just return the mysql client and let the user do whatever they want on top of this

**Comment by servatj** [comment]
Ok thanks a lot, will do that and send you a new PR.

**Pull request merged**